sklearn/linear_model/sag_fast.pyx

```diff
@@ -97,247 +300,264 @@ def sag(SequentialDataset dataset,
+    x_data_ptr : pointer
+        Pointer to the data of the sample x
 
-def get_max_squared_sum(X):
-    """Maximum squared sum of X over samples. 
+    x_ind_ptr : pointer
+        Pointer to the indices of the sample  x
 
-    Used in ``get_auto_step_size()``, for SAG solver.
+    xnnz : int
+        Number of non-zero element in the sample  x
 
-    Parameter
-    ---------
-    X : {numpy array, scipy CSR sparse matrix}, shape (n_samples, n_features)
-        Training vector. X must be in C order.
+    w_data_ptr : pointer
+        Pointer to the data of the weights w
 
-    Returns
-    -------
-    max_squared_sum : double
-        Maximum squared sum of X over samples.
-    """
+    wscale : double
+        Scale of the weights w
 
-    # CSR sparse matrix X
-    cdef np.ndarray[double] X_data
-    cdef np.ndarray[int] X_indptr
-    cdef double *X_data_ptr
-    cdef int *X_indptr_ptr
-    cdef int offset
-
-    # Dense numpy array X
-    cdef np.ndarray[double, ndim=2] X_ndarray
-    cdef int stride
-
-    # Both cases
-    cdef bint sparse = sp.issparse(X)
-    cdef double *x_data_ptr
-    cdef double max_squared_sum = 0.0
-    cdef double current_squared_sum = 0.0
-    cdef int n_samples = X.shape[0]
-    cdef int nnz = X.shape[1]
-    cdef int i, j
-    cdef double val
+    intercept : pointer
+        Pointer to the intercept
 
-    if sparse:
-        X_data = X.data
-        X_indptr = X.indptr
-        X_data_ptr = <double *>X_data.data
-        X_indptr_ptr = <int *>X_indptr.data
-    else:
-        X_ndarray = X
-        stride = X_ndarray.strides[0] / X_ndarray.itemsize
-        x_data_ptr = <double *>X_ndarray.data - stride
+    prediction : pointer
+        Pointer to store the resulting prediction
 
-    with nogil:
-        for i in range(n_samples):
-            # find next sample data
-            if sparse:
-                offset = X_indptr_ptr[i]
-                nnz = X_indptr_ptr[i + 1] - offset
-                x_data_ptr = X_data_ptr + offset
-            else:
-                x_data_ptr += stride
-
-            # sum of squared non-zero features
-            for j in range(nnz):
-                val = x_data_ptr[j]
-                current_squared_sum += val * val
+    n_classes : int
+        Number of classes in multinomial case. Equals 1 in binary case.
 
-            if current_squared_sum > max_squared_sum:
-                max_squared_sum = current_squared_sum
-            current_squared_sum = 0.0
+    """
 
+    cdef int feature_ind, class_ind, j
+    cdef double innerprod
 
-    if not skl_isfinite(max_squared_sum):
-        raise ValueError("Floating-point under-/overflow occurred")
+    for class_ind in range(n_classes):
+        innerprod = 0.0
+        # Compute the dot product only on non-zero elements of x
+        for j in range(xnnz):
+            feature_ind = x_ind_ptr[j]
+            innerprod += (w_data_ptr[feature_ind * n_classes + class_ind] *
+                          x_data_ptr[j])
```

Comment:
So the +11% represents a loss? I thought it was a gain. Then it might be better to revert it back. It might be worth incrementing `w_data_ptr` instead of repeatedly accessing.